diff -durN -x '*~' -x '*.orig' -x '*.rej' jabberd-1.4.3.orig/jabber.xml jabberd-1.4.3/jabber.xml
--- jabberd-1.4.3.orig/jabber.xml	2004-01-03 12:16:10.397562468 +0100
+++ jabberd-1.4.3/jabber.xml	2004-01-03 12:19:00.520106512 +0100
@@ -132,8 +132,11 @@
       notify attribute will send the server administrator(s)
       a message after each valid registration if the notify
       attribute is present.
+      A deny_new attribute will disable registration for new users, but
+      password change and unregistration will still be enabled.
       -->
 
+      <!-- <register deny_new="yes"> -->
       <register notify="yes">
         <instructions>Choose a username and password to register with this server.</instructions>
         <name/>
diff -durN -x '*~' -x '*.orig' -x '*.rej' jabberd-1.4.3.orig/jsm/modules/mod_auth_0k.c jabberd-1.4.3/jsm/modules/mod_auth_0k.c
--- jabberd-1.4.3.orig/jsm/modules/mod_auth_0k.c	2003-10-13 00:58:25.000000000 +0200
+++ jabberd-1.4.3/jsm/modules/mod_auth_0k.c	2004-01-03 12:16:41.492092349 +0100
@@ -226,6 +226,16 @@
     return M_PASS;
 }
 
+mreturn mod_auth_0k_reg_new(mapi m, void *arg)
+{
+    xmlnode reg;
+
+    if((reg = js_config(m->si, "register")) == NULL) return M_PASS;
+    if(xmlnode_get_attrib(reg,"deny_new") != NULL) return M_PASS;
+    return mod_auth_0k_reg(m,arg);
+}
+
+
 /* handle password change requests from a session */
 mreturn mod_auth_0k_server(mapi m, void *arg)
 {
@@ -256,5 +266,5 @@
 
     js_mapi_register(si, e_AUTH, mod_auth_0k_go, enable);
     js_mapi_register(si, e_SERVER, mod_auth_0k_server, NULL);
-    if (js_config(si,"register") != NULL) js_mapi_register(si, e_REGISTER, mod_auth_0k_reg, NULL);
+    js_mapi_register(si, e_REGISTER, mod_auth_0k_reg_new, NULL);
 }
diff -durN -x '*~' -x '*.orig' -x '*.rej' jabberd-1.4.3.orig/jsm/modules/mod_auth_plain.c jabberd-1.4.3/jsm/modules/mod_auth_plain.c
--- jabberd-1.4.3.orig/jsm/modules/mod_auth_plain.c	2002-02-08 08:44:17.000000000 +0100
+++ jabberd-1.4.3/jsm/modules/mod_auth_plain.c	2004-01-03 12:16:41.493092047 +0100
@@ -112,6 +112,15 @@
     return M_PASS;
 }
 
+mreturn mod_auth_plain_reg_new(mapi m, void *arg)
+{
+    xmlnode reg;
+
+    if((reg = js_config(m->si, "register")) == NULL) return M_PASS;
+    if(xmlnode_get_attrib(reg,"deny_new") != NULL) return M_PASS;
+    return mod_auth_plain_reg(m,arg);
+}
+
 /* handle password change requests from a session */
 mreturn mod_auth_plain_server(mapi m, void *arg)
 {
@@ -136,5 +145,5 @@
 
     js_mapi_register(si, e_AUTH, mod_auth_plain_jane, NULL);
     js_mapi_register(si, e_SERVER, mod_auth_plain_server, NULL);
-    if (js_config(si,"register") != NULL) js_mapi_register(si, e_REGISTER, mod_auth_plain_reg, NULL);
+    js_mapi_register(si, e_REGISTER, mod_auth_plain_reg_new, NULL);
 }
diff -durN -x '*~' -x '*.orig' -x '*.rej' jabberd-1.4.3.orig/jsm/modules/mod_register.c jabberd-1.4.3/jsm/modules/mod_register.c
--- jabberd-1.4.3.orig/jsm/modules/mod_register.c	2003-11-08 01:39:44.000000000 +0100
+++ jabberd-1.4.3/jsm/modules/mod_register.c	2004-01-03 12:16:41.493092047 +0100
@@ -45,6 +45,7 @@
     xmlnode reg, x;
 
     if((reg = js_config(m->si, "register")) == NULL) return M_PASS;
+    if(xmlnode_get_attrib(reg,"deny_new") != NULL) return M_PASS;
 
     log_debug("mod_register","checking");
 
